File: scripts/scraper.py
import os
import json
import time
from typing import Union
from requests_html import HTMLSession


BASE_URL = "https://pcpartpicker.com"
PRODUCT_URL = f"{BASE_URL}/products"

ENDPOINTS = [
    "case",
    "case-accessory",
    "case-fan",
    "cpu",
    "cpu-cooler",
    "external-hard-drive",
    "fan-controller",
    "headphones",
    "keyboard",
    "laptop",
    "memory",
    "monitor",
    "motherboard",
    "mouse",
    "os",
    "optical-drive",
    "power-supply",
    "software",
    "sound-card",
    "speakers",
    "internal-hard-drive",
    "thermal-paste",
    "ups",
    "video-card",
    "webcam",
    "wired-network-card",
    "wireless-network-card",
]


def _serialize(value: str) -> Union[str, float]:
    lower = value.lower()

    conversions = {"yes": True, "no": False, "none": None}

    if lower in conversions:
        return conversions[lower]

    return value


def scrape(
    endpoint: str,
    out_dir: str = "./data/raw",
    session: HTMLSession = None,
    page: int = 1,
    paginate: bool = True,
    wait: float = 0.5,
):
    if not session:
        session = HTMLSession()

    res = session.get(f"{PRODUCT_URL}/{endpoint}")
    res.raise_for_status()

    h = res.html
    h.render(sleep=wait)

    max_pages = int(h.find(".pagination > li")[-1].text)
    items = []

    if not os.path.exists(out_dir):
        print(f"Creating {out_dir} because it doesn't exist...\n")
        os.makedirs(out_dir)

    print(f"Scraping /{endpoint}...")

    while page <= max_pages:
        for row in h.find(".tr__product"):
            raw = {}

            raw["name"] = row.find(".td__nameWrapper > p", first=True).text
            raw["rating"] = len(
                row.find(".td__rating", first=True).find(".shape-star-full")
            )
            raw["rating_count"] = int(
                row.find(".td__rating", first=True)
                .text.replace("(", "")
                .replace(")", "")
            )

            price = row.find(".td__price", first=True).text.split("Add")[0]

            if price:
                raw["price_usd"] = float(price[1:])
            else:
                raw["price_usd"] = None

            for col in row.find(".td__spec"):
                split = col.text.split("\n")
                key = split[0].strip().lower().replace(" ", "_")

                if "/" in key:
                    key.replace("/", "per")

                if len(split) == 1:
                    value = None
                else:
                    value = _serialize(split[1])

                raw[key] = value

            items.append(raw)

        print(f"Processed page {page}/{max_pages}")

        if not paginate or page == max_pages:
            break
        else:
            page += 1

    out_path = os.path.join(out_dir, f"{endpoint}.json")

    with open(out_path, "w") as out:
        out.write(json.dumps(items))

    print(f"\nSuccessfully wrote to output file ({out_path})\n")


if __name__ == "__main__":
    session = HTMLSession()

    start = time.time()

    for endpoint in ENDPOINTS:
        scrape(endpoint, session=session)

    end = time.time()

    print(f"Done in {round(end - start, 1)} seconds")

File: src/scraper.ts
import { mkdir, writeFile } from 'fs/promises'
import { join } from 'path'
import type { Page } from 'puppeteer'
import { Cluster } from 'puppeteer-cluster'
import puppeteer from 'puppeteer-extra'
import StealthPlugin from 'puppeteer-extra-plugin-stealth'
import untypedMap from './serialization-map.json'
import {
	customSerializers,
	genericSerialize,
	serializeNumber,
} from './serializers'
import type { Part, PartType, SerializationMap } from './types'

const BASE_URL = 'https://pcpartpicker.com/products'
const STAGING_DIRECTORY = 'data-staging'
const ALL_ENDPOINTS: PartType[] = [
	'cpu',
	'cpu-cooler',
	'motherboard',
	'memory',
	'internal-hard-drive',
	'video-card',
	'case',
	'power-supply',
	'os',
	'monitor',
	'sound-card',
	'wired-network-card',
	'wireless-network-card',
	'headphones',
	'keyboard',
	'mouse',
	'speakers',
	'webcam',
	'case-accessory',
	'case-fan',
	'fan-controller',
	'thermal-paste',
	'external-hard-drive',
	'optical-drive',
	'ups',
]

puppeteer.use(StealthPlugin())

const map = untypedMap as unknown as SerializationMap

async function scrapeInParallel(endpoints: PartType[]) {
	await mkdir(join(STAGING_DIRECTORY, 'json'), { recursive: true })

	const cluster = await Cluster.launch({
		concurrency: Cluster.CONCURRENCY_PAGE,
		maxConcurrency: 5,
		timeout: 1000 * 60 * 20, // 20 minutes
		puppeteer,
		puppeteerOptions: {
			headless: 'new',
		},
	})

	await cluster.task(async ({ page, data: endpoint }) => {
		await page.setViewport({ width: 1920, height: 1080 })

		let fileName = endpoint
		const allParts = []

		try {
			for await (const pageParts of scrape(endpoint, page)) {
				allParts.push(...pageParts)
			}
		} catch (error) {
			console.warn(`[${endpoint}] Aborted unexpectedly:\n\t${error}`)

			if (allParts.length) fileName += '.incomplete'
			else return
		}

		await writeFile(
			join(STAGING_DIRECTORY, 'json', `${fileName}.json`),
			JSON.stringify(allParts)
		)
	})

	cluster.queue('https://pcpartpicker.com', async ({ page, data }) => {
		const res = await page.goto(data)

		try {
			await page.waitForSelector('nav', { timeout: 5000 })
		} catch {
			console.error(
				`Initial fetch test failed (HTTP ${
					res?.status() ?? '?'
				}). Try running with \`{ headless: false }\` to see what the problem is.`
			)
			return
		}

		for (const endpoint of endpoints) {
			cluster.queue(endpoint)
		}
	})

	await cluster.idle()
	await cluster.close()
}

async function* scrape(endpoint: PartType, page: Page): AsyncGenerator<Part[]> {
	await page.setRequestInterception(true)

	page.on('request', (req) => {
		switch (req.resourceType()) {
			case 'font':
			case 'image':
			case 'stylesheet': {
				req.abort()
				break
			}
			default:
				req.continue()
		}
	})

	await page.goto(`${BASE_URL}/${endpoint}`)

	const paginationEl = await page.waitForSelector('.pagination', {
		timeout: 5000,
	})

	// NOTE: We are banging paginationEl because Page.waitForSelector()
	// only returns null when using option `hidden: true`, which we
	// are not using.
	// See: https://pptr.dev/api/puppeteer.page.waitforselector#parameters
	const numPages = await paginationEl!.$eval('li:last-child', (el) =>
		parseInt(el.innerText)
	)

	for (let currentPage = 1; currentPage <= numPages; currentPage++) {
		const pageProducts: Part[] = []

		if (currentPage > 1) {
			await page.goto(`${BASE_URL}/${endpoint}/#page=${currentPage}`)
			await page.waitForNetworkIdle()
		}

		const productEls = await page.$$('.tr__product')

		for (const productEl of productEls) {
			const serialized: Part = {}

			serialized['name'] = await productEl.$eval(
				'.td__name .td__nameWrapper > p',
				(p) => p.innerText.replaceAll('\n', ' ')
			)

			const priceText = await productEl.$eval(
				'.td__price',
				(td) => td.textContent
			)

			if (priceText == null || priceText.trim() === '')
				serialized['price'] = null
			else serialized['price'] = serializeNumber(priceText)

			const specs = await productEl.$$('td.td__spec')

			for (const spec of specs) {
				const specName = await spec.$eval('.specLabel', (l) =>
					(l as HTMLHeadingElement).innerText.trim()
				)
				const mapped = map[endpoint][specName]

				if (typeof mapped === 'undefined')
					throw new Error(`No mapping found for spec '${specName}'`)

				const [snakeSpecName, mappedSpecSerializationType] = mapped

				const specValue = await spec.evaluate(
					(s) => s.childNodes[1]?.textContent
				)

				if (specValue == null || specValue.trim() === '') {
					serialized[snakeSpecName] = null
				} else if (mappedSpecSerializationType === 'custom') {
					serialized[snakeSpecName] =
						customSerializers[endpoint]![snakeSpecName]!(specValue)
				} else {
					serialized[snakeSpecName] = genericSerialize(
						specValue,
						mappedSpecSerializationType
					)
				}
			}

			pageProducts.push(serialized)
		}

		yield pageProducts
	}
}

const inputEndpoints = process.argv.slice(2)
const endpointsToScrape = inputEndpoints.length
	? (inputEndpoints as PartType[])
	: ALL_ENDPOINTS

scrapeInParallel(endpointsToScrape)
